### PR TITLE
[prometheus-pushgateway] update ingress for networking.k8s.io/v1

### LINKED ISSUE #1316: [prometheus-pushgateway] uses deprecated ingress apiVersion
### Describe the bug a clear and concise description of what the bug is.

The chart uses extensions/v1beta1, this got deprecated in kubernetes 1.19 and will be dropped in kubernetes 1.22
Careful, it's more than a version change, there's some more changes required, see https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122

### What's your helm version?

version.BuildInfo{Version:"v3.5.3", GitCommit:"041ce5a2c17a58be0fcd5f5e16fb3e7e95fea622", GitTreeState:"dirty", GoVersion:"go1.15.8"}

### What's your kubectl version?

Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-18T16:12:00Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"} Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.4+k3s1", GitCommit:"3e250fdbab72d88f7e6aae57446023a0567ffc97", GitTreeState:"clean", BuildDate:"2021-08-19T19:09:53Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}

### Which chart?

prometheus-pushgateway

### What's the chart version?

1.10.1

### What happened?

_No response_

### What you expected to happen?

_No response_

### How to reproduce it?

_No response_

### Enter the changed values of values.yaml?

_No response_

### Enter the command that you execute and failing/misfunctioning.

-

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by akloss-cibo**
Signed-off-by: Alec Kloss <akloss@cibotechnologies.com>

@gianrubio@gmail.com
@christian.staude@staffbase.com

#### What this PR does / why we need it:

`networking.k8s.io/v1beta1` is removed in Kubernetes 1.22.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #1316

#### Checklist

- [X] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [X] Chart Version bumped
- [X] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Comment by coryrjohnson** [comment]
@akloss-cibo Can you include `ingressClassName` in the spec for those that might have multiple ingress controller classes? This can also be conditional based on API supporting `networking.k8s.io/v1`

File: charts/prometheus-pushgateway/templates/ingress.yaml [hunk 1]
```diff
@@ -2,7 +2,10 @@
 {{- $serviceName := include "prometheus-pushgateway.fullname" . }}
 {{- $servicePort := .Values.service.port -}}
 {{- $ingressPath := .Values.ingress.path -}}
-{{- if semverCompare ">=1.14.0-0" .Capabilities.KubeVersion.GitVersion }}
+{{- $ingressPathType := .Values.ingress.pathType -}}
```
Comment (helayoty) [inline charts/prometheus-pushgateway/templates/ingress.yaml:5]: values should be `.Values.pushgateway.ingress` , etc..

Comment (akloss-cibo) [reply charts/prometheus-pushgateway/templates/ingress.yaml:5]: What makes you say that?  At least on my helm it works as I would expect as it is here:
```
% helm template --set ingress.enabled=true --set ingress.hosts='{"pushgateway.example.com"}' --set ingress.pathType=foo .  | grep pathType
            pathType: foo
% helm template --set ingress.enabled=true --set ingress.hosts='{"pushgateway.example.com"}' --set pushgateway.ingress.pathType=foo .  | grep pathType
            pathType: ImplementationSpecific
%
```

**Comment by akloss-cibo** [comment]
> @akloss-cibo Can you include `ingressClassName` in the spec for those that might have multiple ingress controller classes? This can also be conditional based on API supporting `networking.k8s.io/v1`

@coryrjohnson added.  I opted to not make it be conditional on `networking.k8s.io/v1`. It violates principle of least astonishment to set something and have it be silently ignored;  I'd rather have something reject the setting if it's used with an ingress version that doesn't support it.  In this case, letting the k8s apiserver reject it is more compatible with don't-repeat-yourself than adding validation into the helm chart repeating the understanding that `ingressClassName` is supported starting in `networking.k8s.io/v1`.

**Comment by akloss-cibo** [comment]
:point_up: forgot the sign-off.

**Review by monotek — CHANGES_REQUESTED** [review]
Imho classname should be configurable too.

You could just use "helm create prometheus-pushgateway" to get a default ingress.yaml and values.yaml which supports it already.

**Comment by akloss-cibo** [comment]
@monotek, I'm not sure what you mean.  Can you clarify?

**Comment by akloss-cibo** [comment]
Is there anything else I can do to move this forward?

**Comment by monotek** [comment]
> @monotek, I'm not sure what you mean. Can you clarify?

If you run the command `helm create prometheus-pushgateway` you'll get a prometheus-pushgateway chart, with a default ingress.yamlm, which has all the stuff implemented.

**Comment by akloss-cibo** [comment]
@monotek ?  Ok, and what should i do with that?

**Comment by monotek** [comment]
> @monotek ? Ok, and what should i do with that?

use it in this pr to be same as suggested helm default.

**Comment by akloss-cibo** [comment]
@monotek The suggested helm default for what?  I'm sorry, I'm having trouble following.  Is there a specific part of this pull request that you think should be changed?

**Comment by akloss-cibo** [comment]
Superseded by https://github.com/prometheus-community/helm-charts/pull/1443

**Pull request closed without merging**